articles/fork-in-C-programming-language/README.md

```diff
@@ -0,0 +1,205 @@
+
+Then the result will be just one "Hello World!".
+
+Now try to execute the following code and compare your result with ours:
+```c
+int doWork(){
+	fork();
+	fork();
+	printf("Hello World!\n");
+}
+int main() {
+	doWork();
+	printf("Hello world!\n");
+	exit(0);
+}
+```
+The result will be:
+```
+Hello world!
+Hello world!
+Hello world!
+Hello world!
+Hello world!
+Hello world!
+Hello world!
+Hello world!
+```
+
+![Fork explaination](/engineering-education/fork-in-C-programming-language/fork2.png)
+
+Because when the process that has been forked inside `dowork()` print `Hello World!` it will continue the main code after the function call and print that `Hello World!` then exit.
+
+### Advanced example 
+When a process creates a new process, then there are two possibilities for the execution exit:
+ - The parent continues to execute concurrently with its child.
+ - The parent waits until some or all of its children have terminated.
+ 
+ ```c
+#include <sys/types.h>
+#include <sys/types.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/wait.h>
+int main(int argc, char *argv[]) {
+ 
+	/* fork a child process */
+	pid_t pid = fork();
+
+	if (pid < 0) { /* error occurred */
+		fprintf(stderr, "Fork Failed");
+		return 1;
+	}
+
+	else if (pid == 0) { /* child process */
+		printf("I'm the child \n"); /* you can execute some commands here */
+	}
+
+	else { /* parent process */
+		/* parent will wait for the child to complete */
+		  wait(NULL);
+		  printf("Child Complete \n");
+	}
+}
+ ```
+The wait call system `wait(NULL)` will make the parent process wait until the child process has executed all of its commands. The result will be:
+```c
+I'm the child 
+Child Complete 
+
+```
+
+Another example:
+ ```c
+#include <sys/types.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/wait.h>
+#include <stdlib.h>
+int main(int argc, char *argv[])
+{
+	printf("I am: %d\n", (int) getpid());
+
+	pid_t pid = fork();
+	printf("fork returned: %d\n", (int) pid);
+
+	if (pid < 0) { /* error occurred */
+		perror("Fork failed");
+	}
+	if (pid == 0) { /* child process */
+		printf("I am the child with pid %d\n", (int) getpid());
+                printf("Child process is exiting\n");
+                exit(0);
+        }
+	/* parent process */
+	printf("I am the parent waiting for the child process to end\n");
+        wait(NULL);
+        printf("parent process is exiting\n");
+        return(0);
+} 
+ ```
+The result will be something like:
+```c
+I am: 2337
+fork returned: 2338
+I am the parent waiting for the child process to end
+fork returned: 0
+I am the child with pid 2338
+Child process is exiting
+parent process is exiting
+```
+
+That's all for today!🥳 
+
+### Conclusion
+We have learned what fork() can do, and how to implement it in the C programming language in unique examples. If you are interested more in the operating system abstractions, and how it is working, then I recommend you to start learning about pipes then semaphores.
```

Comment:
You could include links to recommended reading material (Like articles, documentation, etc)